x-pack/plugins/security_solution/server/lib/telemetry/receiver.ts

```diff
@@ -1320,4 +1349,204 @@ export class TelemetryReceiver implements ITelemetryReceiver {
     }
     return this._esClient;
   }
+
+  public async getIndices(): Promise<string[]> {
+    const es = this.esClient();
+
+    this.logger.l('Fetching indices');
+
+    const request: CatIndicesRequest = {
+      expand_wildcards: ['open', 'hidden'],
+      filter_path: ['index'],
+      format: 'json',
+    };
+
+    return es.cat
+      .indices(request)
+      .then((indices) => indices.map(({ index }) => index as string))
+      .catch((error) => {
+        this.logger.warn('Error fetching indices', { error_message: error } as LogMeta);
+        throw error;
+      });
+  }
+
+  public async getDataStreams(): Promise<DataStream[]> {
+    const es = this.esClient();
+
+    this.logger.l('Fetching datstreams');
+
+    const request: IndicesGetDataStreamRequest = {
+      name: '*',
+      expand_wildcards: ['open', 'hidden'],
+      filter_path: ['data_streams.name', 'data_streams.indices'],
+    };
+
+    return es.indices
+      .getDataStream(request)
+      .then((response) =>
+        response.data_streams.map((ds) => {
+          return {
+            datastream_name: ds.name,
+            indices:
+              ds.indices?.map((index) => {
+                return {
+                  index_name: index.index_name,
+                  ilm_policy: index.ilm_policy,
+                } as Index;
+              }) ?? [],
+          } as DataStream;
+        })
+      )
+      .catch((error) => {
+        this.logger.warn('Error fetching datastreams', { error_message: error } as LogMeta);
+        throw error;
+      });
+  }
+
+  public async *getIndicesStats(indices: string[], config: CommonPrefixesConfig) {
+    const es = this.esClient();
+
+    this.logger.l('Fetching indices stats', { indices } as LogMeta);
+
+    const groupedIndices = findCommonPrefixes(indices, config).map((v, _) => v.parts);
+
+    this.logger.l('Splitted indices into groups', {
+      groups: groupedIndices.length,
+      indices: indices.length,
+    } as LogMeta);
+
+    for (const group of groupedIndices) {
+      for (const name of group) {
+        const request: IndicesStatsRequest = {
+          index: `${name}*`,
+          level: 'indices',
+          metric: ['docs', 'search', 'store'],
+          expand_wildcards: ['open', 'hidden'],
+          filter_path: [
+            'indices.*.total.search.query_total',
+            'indices.*.total.search.query_time_in_millis',
+            'indices.*.total.docs.count',
+            'indices.*.total.docs.deleted',
+            'indices.*.total.store.size_in_bytes',
+          ],
+        };
+
+        try {
+          const response = await es.indices.stats(request);
```

Comment:
I'm concerned about the performance of this collector can we test the following two things:

1. How many API calls would we make as part of this task for clusters where the number of indices lie in the p50 or p95? 
  In our existing telemetry we rather make fewer API calls to collect the information for all the indices we care about.

   We could audit the amount of API calls with a logging config like:
   ```
   logging:
     loggers:
       - name: elasticsearch.query
         level: debug
   ```

2. How long is the event loop is blocked when this task is running on a cluster with an index count at p50 / p95 
  It's difficult to make performance predictions, but the findCommonPrefixes algorithm + all the processing could be quite expensive when clusters have 10s of thousands of indices.

Reply:
Hi @rudolf thanks for your review.
We have run a few performance tests with up to 10K indices. The request that takes more time is `_ilm/policy`, ~4secs. I can share a document with more details about what we did.

On the other hand, be aware that it's possible to configure the job to take the first N indices and data streams, instead of the full list (the default values are 15K indices and 1K data streams, but we can change them).

I'd appreciate if you could help me run further tests against a more representative ES cluster.

Thanks!

Reply:
> the default values are 15K indices and 1K data streams

If we're configuring 15k by default let's test against that. The findCommonPrefixes algorithm would also have worse performance if there's a lot of common prefixes which is not unlike the real world data. So we shouldn't generate index names like `index-01` but rather long ones like `partial-.ds-service-dropwizard-production-metrics-filebeat-7.17.22-aws-us-east-1-2024.11.04-000073` 

Can you share the following results from your test:
 1. how many Elasticsearch API calls are made during the collection
 2. Maximum event loop delay
 3. Maximum RSS
 
You can get all of this by using the following loggers:
```
logging:
  loggers:
    - name: metrics.ops
      level: debug
    - name: elasticsearch.query
      level: debug
ops.interval: 100
logging.appenders.default:
  type: console
  layout:
    type: json
```

Reply:
With these numbers I'm afraid it won't be safe to merge this. On serverless security projects have 2GB RSS for background tasks, but on ECH by default Kibana has 1GB. Even with a 1000 indices this would cause an OOM.

It's possible that the garbage collector is being very lazy and that it's therefore allocating much more heap than what is strictly necessary. To test this locally we'd need to run Kibana with `--max-old-space-size=800` (the amount of heap we allocate on ECH) and ensure it doesn't OOM for the max limit of indices. But locally we would still need to ensure the RSS stays below 1GB because Nodejs allocates `external` memory too.

The event loop delay numbers aren't directly a problem. But if this is what we see on a macbook we would likely see much worse numbers on Cloud where there's a lot more CPU throttling. So ultimately I think we'd need to repeat these tests on Cloud too.